src/os/ObjectStore.h

```diff
@@ -1718,23 +1719,78 @@ class ObjectStore {
    * @param op_flags is CEPH_OSD_OP_FLAG_*
    * @returns number of bytes read on success, or negative error code on failure.
    */
-   virtual int read(
-    const coll_t& cid,
+  virtual int read(
+   const coll_t& cid,
+   const ghobject_t& oid,
+   uint64_t offset,
+   size_t len,
+   bufferlist& bl,
+   uint32_t op_flags = 0) = 0;
+  virtual int read(
+    CollectionHandle &c,
     const ghobject_t& oid,
     uint64_t offset,
     size_t len,
     bufferlist& bl,
-    uint32_t op_flags = 0) = 0;
-   virtual int read(
-     CollectionHandle &c,
-     const ghobject_t& oid,
-     uint64_t offset,
-     size_t len,
-     bufferlist& bl,
-     uint32_t op_flags = 0) {
-     return read(c->get_cid(), oid, offset, len, bl, op_flags);
-   }
+    uint32_t op_flags = 0) {
+    return read(c->get_cid(), oid, offset, len, bl, op_flags);
+  }
+
+  virtual bool has_async_read() const {
+    return false;
+  }
+
+  struct async_read_params_t {
+    uint64_t offset = 0;
+    uint64_t length = 0;
+    ceph::bufferlist* outbl = nullptr;
+    Context* on_complete = nullptr;
+    uint32_t flags = 0;
+
+    async_read_params_t(
+      const uint64_t offset,
+      const uint64_t length,
+      ceph::bufferlist* const outbl,
+      Context* const on_complete,
+      const uint32_t flags)
+    : offset(offset),
+      length(length),
+      outbl(outbl),
+      on_complete(on_complete),
+      flags(flags) {
+    }
+  };
 
+  /**
+   * async_read -- asynchronously read a byte range of data from an object
+   *
+   * Note: if reading from an offset past the end of the object, we
+   * return 0 (not, say, -EINVAL).
+   *
+   * @param cid collection for object
+   * @param oid oid of object
+   * @param offset location offset of first byte to be read
+   * @param len number of bytes to be read
+   * @param bl output bufferlist
+   * @param on_complete callback to be called when data are ready
+   * @param op_flags is CEPH_OSD_OP_FLAG_*
+   * @returns number of bytes read on success, or negative error code on failure.
+   */
+  virtual int async_read(
```

Comment:
I would love to have `async_read()` being asynchronous or not (read: not taking the `PG::lock` twice) depending on its own choice. There are situations, like reading from BS in-memory cache, when the completion phase has little sense. Unfortunately, this would require much deeper changes to the current async infrastructure of OSD. TODO?

Reply:
I think we should definitely design the ObjectStore interface around this idea, even if the OSD can't (yet) make use of it.

What do you think is needed in the OSD to allow it?

Reply:
For data reads we would need to abstract the `pending_async_reads` of `PrimaryLogPG` and move bits like `start_async_reads()` to backends. Also the `do_[sparse_]read` and `objects_read_async` need rework.

OMAP/xattrs is different story as they are entirely synchronous at the moment.

Reply:
FIXED.